packages/tiled/lib/src/tiled_map.dart

```diff
@@ -219,6 +219,36 @@ class TiledMap {
     return imageSet.toList();
   }
 
+  List<TiledImage> collectImagesInLayer(Layer layer) {
+    if (layer is ImageLayer) {
+      return [layer.image];
+    } else if (layer is Group) {
+      return layer.layers.map(collectImagesInLayer).flattened.toList();
+    } else if (layer is TileLayer) {
+      final usedTilesets = <Tileset>{};
+      final inspectedGids = <int>{};
+      const emptyTile = 0;
+      for (final row in layer.tileData ?? <List<Gid>>[]) {
+        for (final gid in row) {
+          final tileGid = gid.tile;
+
+          if (tileGid == emptyTile || inspectedGids.contains(tileGid)) {
+            continue;
+          }
+          inspectedGids.add(tileGid);
+          usedTilesets.add(tilesetByTileGId(tileGid));
+        }
+      }
+
+      return usedTilesets
```

Comment:
I think this could be

```dart
  final rows = layer.tileData ?? <List<Gid>>[];
  final gids = rows.expand((row) => row.map((gid) => gid.tile)).where((gid) => gid != emptyTile).toSet();
  return gids.map(tilesetByTileGId) ...
```

Reply:
I am trying to decide whether or not to divide the cascading in the middle.
Would you kindly give some thoughts on this?
```dart
      return rows
          .expand((row) => row.map((gid) => gid.tile))
          .where((gid) => gid != emptyTile)
          .toSet()
          .map(tilesetByTileGId)
          .toSet()
          .expand(
            (tileset) =>
                [tileset.image, ...tileset.tiles.map((tile) => tile.image)],
          )
          .whereNotNull()
          .toList();
```

```dart
      final gids = rows
          .expand((row) => row.map((gid) => gid.tile))
          .where((gid) => gid != emptyTile)
          .toSet();

      return gids
          .map(tilesetByTileGId)
          .toSet()
          .expand(
            (tileset) =>
                [tileset.image, ...tileset.tiles.map((tile) => tile.image)],
          )
          .whereNotNull()
          .toList();
```

Reply:
The second one might give slightly more readability.
You shouldn't have to do `toSet` again after `map(tilesetByTileGId)`, since the gid list already is unique.

Reply:
> You shouldn't have to do toSet again after map(tilesetByTileGId), since the gid list already is unique.

The gid list is unique but they can be nonunique after `map(tilesetByTileGId)` because different gid can have the same tileset, doesn't it?

Reply:
Ah right, true :)